packages/gator-permissions-snap/src/clients/nonceCaveatClient.ts

```diff
@@ -0,0 +1,99 @@
+import { logger } from '@metamask/7715-permissions-shared/utils';
+import { decodeSingle } from '@metamask/abi-utils';
+import type { Hex } from '@metamask/delegation-core';
+import type { SnapsEthereumProvider } from '@metamask/snaps-sdk';
+
+import { getChainMetadata } from '../core/chainMetadata';
+
+/**
+ * Client that fetches nonce from nonce caveat enforcer.
+ */
+export class NonceCaveatClient {
+  readonly #ethereumProvider: SnapsEthereumProvider;
+
+  // keccak256('currentNonce()')
+  static readonly #currentNonceCalldata = '0x2bd4ed21';
+
+  constructor({
+    ethereumProvider,
+  }: {
+    ethereumProvider: SnapsEthereumProvider;
+  }) {
+    this.#ethereumProvider = ethereumProvider;
+  }
+
+  /**
+   * Fetch the token balance and metadata for a given account and token.
+   * @param args - The parameters for fetching the token balance.
+   * @param args.chainId - The chain ID to fetch the balance from.
+   * @param args.account - The account address to fetch the nonce for.
+   * @returns The nonce.
+   */
+  public async getNonce({
+    chainId,
+    account,
+  }: {
+    chainId: number;
+    account: Hex;
+  }): Promise<bigint> {
+    logger.debug('BlockchainTokenMetadataClient:getTokenBalanceAndMetadata()');
+
+    if (!chainId) {
+      const message = 'No chainId provided to fetch nonce';
+      logger.error(message);
+      throw new Error(message);
+    }
+
+    const { contracts } = getChainMetadata({ chainId });
+
+    if (!account) {
+      const message = 'No account address provided to fetch nonce';
+      logger.error(message);
+      throw new Error(message);
+    }
+
+    // Check if we're on the correct chain
```

Comment:
In https://github.com/MetaMask/snap-7715-permissions/blob/271371016a70afa84f818fa5c1e1d3d6000dda23/packages/gator-permissions-snap/src/core/accountController.ts#L145 I introduce a better way to handle this - if the chain doesn't match, we request the wallet switch (the currently selected chain _for the snap_) before checking again to ensure that it's switched successfully.

Although that PR is not ready, it might be worthwhile incorporating that same approach here.

Maybe it's something we will want to abstract away too - perhaps we can encapsulate the SnapsEthereumProvider in a `ChainSwitchingSnapsEthereumProvider` 🤷